BloomFilter.md

```diff
@@ -49,87 +44,185 @@ pushdown even in columns with high cardinality and when space is at a premium.
+operate on. It then uses the least significant 32 bits of the argument
+to `filter_insert` as an argument to `block_insert` called on that
+block.
+
+The `filter_check` operation uses the same method as `filter_insert`
+to select a block to operate on, then uses the least significant 32
+bits of its argument as an argument to `block_check` called on that
+block, returning the result.
+
+In the pseudocode below, the modulus operator is represented with the C
+language's "`%`" operator. The "`>>`" operator is used to denote the
+conversion of an unsigned 64-bit integer to an unsigned 32-bit integer
+containing only the most significant 32 bits, and C's cast operator
+"`(unsigned int32)`" is used to denote the conversion of an unsigned
+64-bit integer to an unsigned 32-bit integer containing only the lest
+significant 32 bits.
+
+```
+void filter_insert(SBBF filter, unsigned int64 x) {
+  block b = filter.getBlock((x >> 32) % filter.numberOfBlocks());
+  block_insert(b, (unsigned int32)x)
 }
+```
 
 ```
+boolean filter_check(SBBF filter, unsigned int64 x) {
+  block b = filter.getBlock((x >> 32) % filter.numberOfBlocks());
+  return block_check(b, (unsigned int32)x)
+}
+```
+
+The use of blocks is from Putze et al.'s [Cache-, Hash- and
+Space-Efficient Bloom
+filters](http://algo2.iti.kit.edu/documents/cacheefficientbloomfilters-jea.pdf)
 
-#### Hash Function
-The function used to hash values in the initial implementation is
-[xxHash](https://cyan4973.github.io/xxHash/), using the function XXH64 with a
-seed of 0 and [following the specification version
+To use an SBBF for values of arbitrary Parquet types, we apply a hash
+function to that value - at the time of writing,
+[xxHash](https://cyan4973.github.io/xxHash/), using the function XXH64
+with a seed of 0 and [following the specification version
 0.1.1](https://github.com/Cyan4973/xxHash/blob/v0.7.0/doc/xxhash_spec.md).
 
-#### Build a Bloom filter
-The fact that exactly eight bits are checked during each lookup means that these filters
-are most space efficient when used with an expected false positive rate of about
-0.5%. This is achieved when there are about 11.54 bits for every distinct value inserted
-into the filter.
+#### Sizing an SBBF
 
-To calculate the size the filter should be for another false positive rate `p`, use the
-following formula. The output is in bits per distinct element:
+The `check` operation in SBBFs can return `true` for an argument that
+was never inserted into the SBBF. These are called "false
+positives". There is not a simple closed-form calculation of this
+probability, but empirically, an SBBF that uses 6 bits of space (or
+6/256 blocks) per unique item inserted will have a false positive
+probability of around 10%.
 
-```c
-c = -8 / log(1 - pow(p, 1.0 / 8))
-```
+For example, consider an SBBF that contains 1024 blocks and has been
+loaded with (256/6) * 1024 = 43691 unique keys. Running `filter_check`
```

Comment:
I think it would be more clear to add units to this example and group the 1024 blocks * 256 bits/block to get the total number of bits first, then divide by bits per key.

It sounds like the way to size is to estimate the number of expected values per row group, say 10,000 is:
* Choose the desired false-positive rate, like 1% and multiply by values to get total required bits: 10.5 bits * 10,000 values = 105,000 bits.
* Divide total bits by 256 to get the required number of blocks: 105,000 bits / 256 bits per block = 410 blocks.
* Find the first power of 2 higher than the number of blocks: 2^8 = 256 < 410, 2^9 = 512 > 410, so use 512 blocks.

Is this correct?

Reply:
Yes, correct.